README.md
[![PyPI version](https://img.shields.io/pypi/v/pygellermann.svg)](https://pypi.python.org/pypi/pygellermann)
[![PyPI Python versions](https://img.shields.io/pypi/pyversions/pygellermann.svg)](https://pypi.python.org/pypi/pygellermann)
[![GitHub Actions status](https://github.com/YannickJadoul/PyGellermann/actions/workflows/ci.yml/badge.svg)](https://github.com/YannickJadoul/PyGellermann/actions/workflows/ci.yml)
[![pre-commit.ci status](https://results.pre-commit.ci/badge/github/YannickJadoul/PyGellermann/main.svg)](https://results.pre-commit.ci/latest/github/YannickJadoul/PyGellermann/main)

# PyGellerman: A Python Gellermann series generator

This is a Python implementation of a random genenerator for Gellermann series, pseudorandom binary sequences for human and non-human animal behavioural experiments[^Gellermann1933]. It includes both a graphical user interface (GUI) as well as a simple Python API.

Gellermann series aim to avoid inflating a partipant's experimental performance by excluding random sequences that would match simple psychological or behavioural patterns. More specifically, a Gellermann series is a random sequence which satisfies five criteria; each series of length $n$:
- must contain an equal number (= $n/2$) of As and Bs;
- must contain at most 3 As or Bs in a row;
- must contain at least 20% (= $n/5$) As and Bs within both the first and last
half;
- must contain at most $n/2$ reversals (A-B or B-A transitions);
- must provide a correct response rate close to 50% chance when responses are provided as simple alternation (ABAB...) or double alternation (AABBAA... and ABBAAB...).

A full, open access manuscript introducing PyGellermann and describing its scientific applications and limitations has been published in BMC Research Notes, and is available [here](https://bmcresnotes.biomedcentral.com/articles/10.1186/s13104-023-06396-x). It can be cited as follows:

> Jadoul, Y., Duengen, D., & Ravignani, A. (2023). PyGellermann: a Python tool to generate pseudorandom series for human and non-human animal behavioural experiments. *BMC Research Notes, 16*(1), 135. https://doi.org/10.1186/s13104-023-06396-x

## Installation
PyGellermann is available on PyPI and can be installed using pip:

    pip install pygellermann

For details on how to use `pip`, see the [Python Packaging User Guide](https://packaging.python.org/tutorials/installing-packages/) or [pip's User Guide](https://pip.pypa.io/en/stable/user_guide/).

Alternatively, you can download the GUI as a standalone executable for Linux, macOS, and Windows from the [releases page](https://github.com/YannickJadoul/PyGellermann/releases).

**Note**: On macOS, opening the PyGellermann app bundle might at first be blocked because it is not signed or distributed through the App Store. To override this and open the app, right-click on the app bundle and select *Open* from the context menu. This will open a dialog asking you to confirm that you want to open the app. After that, you can open the app by double-clicking on it. For more details, see [the macOS User Guide](https://support.apple.com/guide/mac-help/open-a-mac-app-from-an-unidentified-developer-mh40616/mac).

## Usage
### Graphical User Interface
After installation, you can run the GUI by typing `pygellermann-gui` in your terminal or by running the standalone executable (`PyGellermann`, `PyGellermann.app`, or `PyGellermann.exe`). The following window should appear:

![PyGellermann GUI](https://raw.githubusercontent.com/YannickJadoul/PyGellermann/main/docs/gui.png)

Four parameters can be set to customize the generated Gellermann series:
- *Sequence length*: The length of each generated sequence.
- *Number of sequences*: The number of sequences to generate.
- *Alternation tolerance*: How close to 50% chance level a sequence needs to be when compared to single or double alternation.
- *Choices*: The two possible choices for each stimulus in the sequences.
- *Random seed*: The seed for the random number generator, allowing to deterministically generate the same sequences.

The *Generate* button will then generate the requested number of sequences, and display them in the table underneath.

Finally, the generated Gellermann series can be copied to the clipboard (*Copy*) or saved to a CSV file (*Save...*).

### Python API
The Python API consists of 4 simple functions:

- `is_gellermann_series(s, alternation_tolerance=DEFAULT_ALTERNATION_TOLERANCE)`

  Check if a binary sequence is a Gellermann series.

  #### Parameters
  - `s` : `Sequence[Any]`

    A binary series (i.e., containing two different elements) of even length.

  - `alternation_tolerance` : `float`, optional

    The tolerance around 50% chance level compared to single or double alternation, a value between 0 and 0.5 (default: 0.1).

  #### Returns
  - `bool`

    True if the given sequence is a Gellermann series, False otherwise.

  #### Raises
  - `ValueError`

    If the sequence length is not even, or if the sequence contains more than two different elements, or if the alternation tolerance is not between 0 and 0.5.

  #### Examples
  ```pycon
  >>> is_gellermann_series(['B', 'B', 'A', 'B', 'A', 'B', 'B', 'A', 'A', 'A'])
  True
  >>> is_gellermann_series('1112212122122211', alternation_tolerance=0.2)
  True
  >>> is_gellermann_series('1112212122122211', alternation_tolerance=0.0)
  False
  ```

- `generate_gellermann_series(n, m, choices=('A', 'B'), rng=None, **kwargs)`

  Generate m random Gellermann series of length n.

  #### Parameters
  - `n` : `int`

    The length of the series.

  - `m` : `int`

    The number of series to generate.

  - `choices` : `Tuple[Any, Any]`, optional

    The two elements of the series (default: ('A', 'B')).

  - `rng` : `np.random.Generator`, optional

    A NumPy random number generator (default: `None`, which uses the default NumPy random number generator).

  - `max_iterations` : `int`, optional

    The maximum number of iterations to try to generate all Gellermann series (default: `None`, which tries indefinitely).

  - `kwargs`

    Additional keyword arguments passed to `is_gellermann_series`.

  #### Yields
  - `Iterator[Sequence[Any]]`

    A generator object with m Gellermann series of length n.


- `generate_all_gellermann_series(n, choices, **kwargs)`
  Generate all Gellermann series of length n in lexicographic order.

  #### Parameters
  - `n` : `int`

    The length of the series.

  - `choices` : `Tuple[Any, Any]`, optional

      The two elements of the series (default: ('A', 'B')).

  - `kwargs`

      Additional keyword arguments passed to `is_gellermann_series`.

  #### Yields
  - `Iterator[Sequence[Any]]`

    A generator object with all Gellermann series of length n.


- `generate_gellermann_series_table(n, m, long_format: bool = False, **kwargs)`

  Generate a Pandas DataFrame of m random Gellermann series of length n.

  In the wide format, the DataFrame has columns 'series_i', 'element_0', 'element_1', ..., 'element_{n-1}', and each row contains a full series. In the long format, the DataFrame has columns 'series_i', 'element_i', 'element', and each row contains a single element of a series.

  #### Parameters
  - `n` : `int`

    The length of the series.

  - `m` : `int`

    The number of series to generate.

  - `long_format` : `bool`, optional

    If True, the DataFrame is in long format (default: False).

  - `kwargs`

    Additional keyword arguments passed to `generate_gellermann_series`.

  #### Returns
  - `pd.DataFrame`

    A Pandas DataFrame of m random Gellermann series of length n.


## License

PyGellermann is released under the GNU General Public License v3 or later. See the LICENSE file for details.

PyGellermann was developed at the [Comparative Bioacoustics Group](https://www.mpi.nl/department/comparative-bioacoustics/20) of the [Max Planck Institute for Psycholinguistics](https://www.mpi.nl/), Nijmegen, the Netherlands.

In scientific publications, PyGellermann can be cited as follows:

> Jadoul, Y., Duengen, D., & Ravignani, A. (2023). PyGellermann: a Python tool to generate pseudorandom series for human and non-human animal behavioural experiments. *BMC Research Notes, 16*(1), 135. https://doi.org/10.1186/s13104-023-06396-x

## References

[^Gellermann1933]: Gellermann, L. W. (1933). Chance orders of alternating stimuli in visual discrimination experiments. *The Journal of Genetic Psychology, 42*, 206-208.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
_Nothing yet_

## [0.1.0] - 2023-03-02
### Added
- Initial Python API, with `is_gellermann_series`, `generate_gellermann_series`, `generate_all_gellermann_series`, and `generate_gellermann_series_table`
- Basic GUI, with PyInstaller-generated standalone executables for Linux, macOS, and Windows

[Unreleased]: https://github.com/YannickJadoul/PyGellermann/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/YannickJadoul/PyGellermann/compare/e013a30...v0.1.0


tests/test_gellermann.py
# Copyright (C) 2022-2023  Yannick Jadoul
#
# This file is part of PyGellermann.
#
# PyGellermann is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# PyGellermann is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with PyGellermann.  If not, see <https://www.gnu.org/licenses/>.


import pytest

import pygellermann

import numpy as np


def test_is_gellermann_series():
    assert pygellermann.is_gellermann_series("R R R L L R L R L L".split())
    assert pygellermann.is_gellermann_series("L L L R R L R L R R".split())
    assert pygellermann.is_gellermann_series("R R L R L R R L L L".split())
    assert pygellermann.is_gellermann_series("L L R L R L L R R R".split())
    assert pygellermann.is_gellermann_series("R R L L R L R R L L".split())
    assert pygellermann.is_gellermann_series("L L R R L R L L R R".split())

    assert pygellermann.is_gellermann_series("LLRRLRLLRR")
    assert pygellermann.is_gellermann_series("LLRRLRLLRR".replace("L", "A").replace("R", "B"))


@pytest.mark.parametrize('m', [1, 5, 20])
@pytest.mark.parametrize('n', [10, 20, 40])
def test_generate_gellermann_series(n, m):
    all_series = list(pygellermann.generate_gellermann_series(n, m))
    assert len(all_series) == m
    assert all(len(s) == n for s in all_series)
    assert all(pygellermann.is_gellermann_series(s) for s in all_series)
    assert set.union(*map(set, all_series)) == {'A', 'B'}


def test_generate_gellermann_series_rng():
    series = list(pygellermann.generate_gellermann_series(10, 5, rng=np.random.default_rng(42)))
    assert series == [
        ['A', 'A', 'A', 'B', 'B', 'A', 'B', 'A', 'B', 'B'],
        ['B', 'B', 'A', 'A', 'B', 'A', 'B', 'B', 'A', 'A'],
        ['A', 'A', 'A', 'B', 'B', 'A', 'B', 'A', 'B', 'B'],
        ['B', 'B', 'A', 'A', 'B', 'A', 'B', 'B', 'A', 'A'],
        ['B', 'B', 'A', 'A', 'B', 'A', 'B', 'B', 'A', 'A']
    ]


def test_generate_gellermann_series_max_iterations():
    series = list(pygellermann.generate_gellermann_series(8, 5, max_iterations=1000))
    assert len(series) == 0


@pytest.mark.parametrize('n, m_expected', [(10, 8), (16, 80), (20, 4726)])
def test_generate_all_gellermann_series(n, m_expected):
    all_series = list(pygellermann.generate_all_gellermann_series(n))
    assert len(all_series) == m_expected
    assert all(len(s) == n for s in all_series)
    assert set.union(*map(set, all_series)) == {'A', 'B'}

    for s in pygellermann.generate_gellermann_series(n, 100):
        assert pygellermann.is_gellermann_series(s) == (s in all_series)


@pytest.mark.parametrize('tolerance, m_expected', [(0.1, 8), (0.2, 32), (0.3, 72), (0.4, 84), (0.5, 86)])
def test_generate_all_gellermann_series_tolerance(tolerance, m_expected):
    all_series = list(pygellermann.generate_all_gellermann_series(10, alternation_tolerance=tolerance))
    assert len(all_series) == m_expected
    assert all(len(s) == 10 for s in all_series)
    assert set.union(*map(set, all_series)) == {'A', 'B'}


@pytest.mark.parametrize('choices', [('L', 'R'), ('R', 'L'), (1, 2), ('ABC', (42,))])
@pytest.mark.parametrize('n', [10, 20, 40])
def test_generate_gellermann_series_choices(n, choices):
    series = next(pygellermann.generate_gellermann_series(n, 1, choices=choices))
    assert pygellermann.is_gellermann_series(series)
    assert len(series) == n
    assert set(series) == set(choices)


@pytest.mark.parametrize('m', [1, 5, 20])
@pytest.mark.parametrize('n', [10, 20, 40])
def test_generate_gellermann_series_table_wide(n, m):
    df = pygellermann.generate_gellermann_series_table(n, m, long_format=False)

    assert len(df) == m
    assert len(df.columns) == n
    assert list(df.columns) == [f'element_{i}' for i in range(n)]
    assert df.index.nlevels == 1
    assert df.index.names == ['series_i']
    assert all(pygellermann.is_gellermann_series(r) for _, r in df.iterrows())


@pytest.mark.parametrize('m', [1, 5, 20])
@pytest.mark.parametrize('n', [10, 20, 40])
def test_generate_gellermann_series_table_long(n, m):
    df = pygellermann.generate_gellermann_series_table(n, m, long_format=True)

    assert len(df) == m * n
    assert len(df.columns) == 1
    assert list(df.columns) == ['element']
    assert df.index.nlevels == 2
    assert df.index.names == ['series_i', 'element_i']
    assert all(pygellermann.is_gellermann_series(g['element']) for _, g in df.groupby('series_i'))


pygellermann/gui.py
# Copyright (C) 2022-2023  Yannick Jadoul
#
# This file is part of PyGellermann.
#
# PyGellermann is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# PyGellermann is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with PyGellermann.  If not, see <https://www.gnu.org/licenses/>.

"""GUI wrapping functionality for generating Gellermann series."""


from qtpy import QtCore, QtGui, QtWidgets

from . import __version__, gellermann

import numpy as np
import qdarktheme  # type: ignore

import sys


class ShorterLineEdit(QtWidgets.QLineEdit):
    def sizeHint(self):
        return QtCore.QSize(20, super().sizeHint().height())


class MainWindow(QtWidgets.QWidget):

    def __init__(self):
        super().__init__()

        self._series = []
        self._was_saved = False

        self.setWindowTitle(f"PyGellermann {__version__}")
        self.setMinimumWidth(500)

        self._sequence_length_spinbox = QtWidgets.QSpinBox()
        self._sequence_length_spinbox.setRange(2, 100)
        self._sequence_length_spinbox.setSingleStep(2)
        self._sequence_length_spinbox.setValue(10)
        self._number_of_sequences_spinbox = QtWidgets.QSpinBox()
        self._number_of_sequences_spinbox.setRange(1, 1000)
        self._number_of_sequences_spinbox.setValue(5)
        self._alternation_tolerance_spinbox = QtWidgets.QDoubleSpinBox()
        self._alternation_tolerance_spinbox.setRange(0.0, 0.5)
        self._alternation_tolerance_spinbox.setSingleStep(0.01)
        self._alternation_tolerance_spinbox.setValue(0.1)

        self._option1_lineedit = ShorterLineEdit()
        self._option1_lineedit.setText("A")
        self._option2_lineedit = ShorterLineEdit()
        self._option2_lineedit.setText("B")

        random_seed_validator = QtGui.QIntValidator()
        random_seed_validator.setBottom(0)
        self._random_seed_lineedit = ShorterLineEdit()
        self._random_seed_lineedit.setValidator(random_seed_validator)
        self._random_seed_lineedit.setPlaceholderText("None")

        generate_button = QtWidgets.QPushButton("Generate")
        generate_button.clicked.connect(self._generate)

        self._results_table = QtWidgets.QTableWidget()
        self._results_table.setSelectionBehavior(QtWidgets.QAbstractItemView.SelectionBehavior.SelectRows)
        self._results_table.horizontalHeader().setVisible(False)
        self._results_table.setMinimumHeight(200)

        self._results_widget = QtWidgets.QWidget()
        self._results_widget.setVisible(False)

        clear_button = QtWidgets.QPushButton("Clear")
        clear_button.clicked.connect(self._clear)
        copy_button = QtWidgets.QPushButton("&Copy")
        copy_button.clicked.connect(self._copy)
        copy_button.setShortcut(QtGui.QKeySequence.StandardKey.Copy)
        save_button = QtWidgets.QPushButton("&Save...")
        save_button.clicked.connect(self._save)
        save_button.setShortcut(QtGui.QKeySequence.StandardKey.Save)
        self._selected_format = ""

        input_layout = QtWidgets.QFormLayout()
        input_layout.addRow("Sequence &length:", self._sequence_length_spinbox)
        input_layout.addRow("&Number of sequences:", self._number_of_sequences_spinbox)
        input_layout.addRow("&Alternation tolerance:", self._alternation_tolerance_spinbox)
        input_layout.addRow("Choices:", self._option1_lineedit)
        input_layout.addRow("", self._option2_lineedit)
        input_layout.addRow("Random seed:", self._random_seed_lineedit)
        input_layout.addRow(generate_button)

        centered_input_layout = QtWidgets.QHBoxLayout()
        centered_input_layout.addStretch()
        centered_input_layout.addLayout(input_layout)
        centered_input_layout.addStretch()

        button_layout = QtWidgets.QHBoxLayout()
        button_layout.addStretch()
        button_layout.addWidget(clear_button)
        button_layout.addWidget(copy_button)
        button_layout.addWidget(save_button)

        results_layout = QtWidgets.QVBoxLayout()
        results_layout.addWidget(self._results_table)
        results_layout.addLayout(button_layout)
        results_layout.setContentsMargins(0, 0, 0, 0)
        self._results_widget.setLayout(results_layout)

        main_layout = QtWidgets.QVBoxLayout()
        main_layout.addLayout(centered_input_layout)
        main_layout.addWidget(self._results_widget)

        self.setLayout(main_layout)

        self._collapse_results()

    def closeEvent(self, event):
        if not self._check_unsaved_changes():
            event.ignore()

    def _check_unsaved_changes(self):
        if not self._series or self._was_saved:
            return True

        StandardButton = QtWidgets.QMessageBox.StandardButton
        message = "These generated Gellermann series have not been saved yet. Do you want to save them?"
        reply = QtWidgets.QMessageBox.question(self, "Unsaved changes", message, StandardButton.Yes | StandardButton.No | StandardButton.Cancel)
        if reply == StandardButton.Yes:
            return self._save()
        elif reply == StandardButton.No:
            return True
        else:
            return False

    def _generate(self):
        if not self._check_unsaved_changes():
            return

        n = self._sequence_length_spinbox.value()
        m = self._number_of_sequences_spinbox.value()
        alternation_tolerance = self._alternation_tolerance_spinbox.value()
        choices = (self._option1_lineedit.text(), self._option2_lineedit.text())

        if n % 2:
            QtWidgets.QMessageBox.critical(self, "Error", "PyGellermann cannot yet generate odd length Gellermann series.")
            return

        try:
            random_seed_text = self._random_seed_lineedit.text()
            random_seed = int(random_seed_text) if random_seed_text else None
        except ValueError:
            QtWidgets.QMessageBox.critical(self, "Error", "Invalid random seed.")
            return
        rng = np.random.default_rng(random_seed)

        progress_dialog = QtWidgets.QProgressDialog("Generating series...", "Cancel", 0, m, self)
        progress_dialog.setWindowTitle("PyGellermann")
        progress_dialog.setMinimumDuration(500)
        progress_dialog.setModal(True)

        self._results_table.clearContents()
        self._results_table.setRowCount(0)

        self._series = []
        try:
            while len(self._series) < m:
                s = next(gellermann.generate_gellermann_series(n, 1, alternation_tolerance=alternation_tolerance,
                                                               choices=choices, rng=rng, max_iterations=1000), None)

                if progress_dialog.wasCanceled():
                    break

                if s is not None:
                    self._series.append(s)

                progress_dialog.setValue(len(self._series))
                QtWidgets.QApplication.processEvents()
        except Exception as e:
            QtWidgets.QMessageBox.critical(self, "Error", f"An unexpected error occurred: {e}")
            progress_dialog.cancel()
            progress_dialog.close()

        if progress_dialog.wasCanceled():
            self._series = []
            self._collapse_results()
            return

        self._results_table.setRowCount(len(self._series))
        self._results_table.setColumnCount(len(self._series[0]))
        for i, row in enumerate(self._series):
            for j, element in enumerate(row):
                item = QtWidgets.QTableWidgetItem(element)
                item.setFlags(item.flags() & ~QtCore.Qt.ItemIsEditable)  # type: ignore
                self._results_table.setItem(i, j, item)
        self._results_table.setHorizontalHeaderLabels([""] * self._results_table.columnCount())
        min_width = (max(self._results_table.sizeHintForColumn(i) for i in range(self._results_table.columnCount())))
        self._results_table.horizontalHeader().setMinimumSectionSize(min_width)
        self._results_table.resizeColumnsToContents()
        self._results_table.resizeRowsToContents()
        self._expand_results()

        self._was_saved = False

    def _copy(self):
        row_strings = []
        for row in range(self._results_table.rowCount()):
            row_strings.append("\t".join(self._results_table.item(row, column).text() for column in range(self._results_table.columnCount())))
        QtWidgets.QApplication.clipboard().setText("\n".join(row_strings))

    def _clear(self):
        if not self._check_unsaved_changes():
            return

        self._series = []
        self._results_table.clearContents()
        self._results_table.setRowCount(0)
        self._collapse_results()

    def _collapse_results(self):
        self.installEventFilter(self)
        self._results_widget.setVisible(False)
        self.setMinimumHeight(self.sizeHint().height())
        QtCore.QTimer.singleShot(20, self.adjustSize)

    def _expand_results(self):
        self._results_widget.setVisible(True)
        self.setMinimumHeight(self.sizeHint().height())

    def _save(self):
        formats = [
            "Comma-separated file, wide-format (*.csv)",
            "Comma-separated file, long-format (*.csv)"
        ]
        name, selected = QtWidgets.QFileDialog.getSaveFileName(self, "Save generated series to file", "", ";;".join(formats), self._selected_format)

        if not name:
            return False
        self._selected_format = selected

        if selected == formats[0]:
            df = gellermann._series_to_wide_format_df(self._series)
        else:
            df = gellermann._series_to_long_format_df(self._series)

        try:
            df.to_csv(name)
        except Exception as e:
            QtWidgets.QMessageBox.critical(self, "Error", f"Could not save series to file:\n\n{e}")
            return False

        self._was_saved = True

        return True


def main():
    import signal
    signal.signal(signal.SIGINT, signal.SIG_DFL)

    qdarktheme.enable_hi_dpi()

    app = QtWidgets.QApplication(sys.argv)
    qdarktheme.setup_theme('auto')
    main_window = MainWindow()
    main_window.show()
    sys.exit(app.exec_())  # type: ignore


if __name__ == '__main__':
    main()


pygellermann/gellermann.py
# Copyright (C) 2022-2023  Yannick Jadoul
#
# This file is part of PyGellermann.
#
# PyGellermann is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# PyGellermann is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with PyGellermann.  If not, see <https://www.gnu.org/licenses/>.

"""Functionality for checking and generating Gellermann series."""

import numpy as np
import pandas as pd

import itertools
import sys

import numpy.typing as npt
from typing import Any, Iterator, List, Optional, Sequence, Tuple

# TODO: uneven length sequences
# TODO: optimize multiple calculations
# TODO: C backend?

DEFAULT_ALTERNATION_TOLERANCE = 0.1


if sys.version_info >= (3, 9):
    BoolSequence = np.ndarray[int, np.dtype[np.bool_]]
else:
    BoolSequence = npt.NDArray[np.bool_]


def balanced_elements(s: BoolSequence) -> bool:
    """Check if a boolean sequence has equal number of True and False elements."""
    n = len(s)
    return n % 2 == 0 and np.sum(s) == n // 2


def more_than_three_successive(s: BoolSequence) -> bool:
    """Check if a boolean sequence has more than three True or False elements in a row."""
    successive = s[:-1] == s[1:]
    return bool(np.any(successive[:-2] & successive[1:-1] & successive[2:]))


def at_least_twenty_percent_per_half(s: BoolSequence) -> bool:
    """Check if a boolean sequence has at least 20% True or False elements in both halves."""
    n = len(s)
    p = n // 5
    return bool(np.sum(s[:n//2]) >= p and np.sum(~s[:n//2]) >= p and np.sum(s[n//2:]) >= p and np.sum(~s[n//2:]) >= p)


def less_than_half_reversals(s: BoolSequence) -> bool:
    """Check if a boolean sequence has less than half True-False or False-True reversals."""
    return int(np.sum(s[:-1] != s[1:])) <= len(s) // 2


def close_to_fifty_percent_alternation(s: BoolSequence, tolerance: float) -> bool:
    """Check if a boolean sequence matches the single or double alternation criterion.

    Checks if matches single or double alternating sequences sufficiently close to 50% chance
    level, +/- tolerance.
    """
    assert 0 <= tolerance <= 0.5

    n = len(s)

    def close_to_fifty_percent_for(alternation: BoolSequence) -> bool:
        return 0.5 - tolerance <= int(np.sum(alternation == s)) / n <= 0.5 + tolerance

    return (close_to_fifty_percent_for(np.tile([True, False], (n + 1) // 2)[:n]) and
            close_to_fifty_percent_for(np.tile([True, True, False, False], (n + 3) // 4)[:n]) and
            close_to_fifty_percent_for(np.tile([True, False, False, True], (n + 3) // 4)[:n]))


def is_boolean_gellermann_series(s: BoolSequence, alternation_tolerance: float = DEFAULT_ALTERNATION_TOLERANCE) -> bool:
    """Check if a boolean sequence is a Gellermann series."""
    assert len(s) % 2 == 0
    assert 0 <= alternation_tolerance <= 0.5

    return (balanced_elements(s) and
            not more_than_three_successive(s) and
            at_least_twenty_percent_per_half(s) and
            less_than_half_reversals(s) and
            close_to_fifty_percent_alternation(s, alternation_tolerance))


def is_gellermann_series(s: Sequence[Any], alternation_tolerance: float = DEFAULT_ALTERNATION_TOLERANCE) -> bool:
    """Check if a binary sequence is a Gellermann series.

    Parameters
    ----------
    s
        A binary series (i.e., containing two different elements) of even length.
    alternation_tolerance
        The tolerance around 50% chance level compared to single or double alternation, a value
        between 0 and 0.5 (default: 0.1).

    Returns
    -------
    bool
        True if the given sequence is a Gellermann series, False otherwise.

    Raises
    ------
    ValueError
        If the sequence length is not even, or if the sequence contains more than two different
        elements, or if the alternation tolerance is not between 0 and 0.5.

    Examples
    --------
    >>> is_gellermann_series(['B', 'B', 'A', 'B', 'A', 'B', 'B', 'A', 'A', 'A'])
    True
    >>> is_gellermann_series('1112212122122211', alternation_tolerance=0.2)
    True
    >>> is_gellermann_series('1112212122122211', alternation_tolerance=0.0)
    False
    """
    if len(s) % 2 != 0:
        raise ValueError(f"Sequence length {len(s)} is not even.")
    if len(set(s)) > 2:
        raise ValueError(f"Sequence {s} contains more than two different elements.")
    if not 0 <= alternation_tolerance <= 0.5:
        raise ValueError(f"Alternation tolerance {alternation_tolerance} is not between 0 and 0.5.")

    if len(s) == 0:
        return True

    s = list(s)
    return is_boolean_gellermann_series(np.array([x == s[0] for x in s]), alternation_tolerance=alternation_tolerance)


def generate_boolean_gellermann_series(n: int, m: int, rng: Optional[np.random.Generator] = None,
                                       max_iterations: Optional[int] = None, **kwargs: Any) -> Iterator[BoolSequence]:
    """Generate m random boolean Gellermann series of length n."""
    assert n % 2 == 0
    assert m > 0

    if rng is None:
        rng = np.random.default_rng()
    s = np.repeat([True, False], n // 2)

    for _ in itertools.islice(itertools.count(), max_iterations):
        rng.shuffle(s)
        if is_boolean_gellermann_series(s, **kwargs):
            yield s.copy()
            m -= 1

        if m == 0:
            break


def generate_gellermann_series(n: int, m: int, choices: Tuple[Any, Any] = ('A', 'B'), rng: Optional[np.random.Generator] = None,
                               max_iterations: Optional[int] = None, **kwargs: Any) -> Iterator[Sequence[Any]]:
    """Generate m random Gellermann series of length n.

    Parameters
    ----------
    n
        The length of the series.
    m
        The number of series to generate.
    choices
        The two elements of the series (default: ('A', 'B')).
    rng
        A NumPy random number generator (default: None, which uses the default NumPy random number
        generator).
    max_iterations
        The maximum number of iterations to try to generate all Gellermann series (default: None,
        which tries indefinitely).
    kwargs
        Additional keyword arguments passed to `is_gellermann_series`.

    Yields
    ------
    Iterator[Sequence[Any]]
        A generator object with m Gellermann series of length n.
    """
    for s in generate_boolean_gellermann_series(n, m, rng=rng, max_iterations=max_iterations, **kwargs):
        yield [choices[int(x)] for x in s]


def generate_all_boolean_gellermann_series(n: int, **kwargs: Any) -> Iterator[BoolSequence]:
    """Generate all boolean Gellermann series of length n."""
    assert n % 2 == 0

    for s in itertools.product([False, True], repeat=n):
        if is_boolean_gellermann_series(np.array(s), **kwargs):
            yield np.array(s)


def generate_all_gellermann_series(n: int, choices: Tuple[Any, Any] = ('A', 'B'), **kwargs: Any) -> Iterator[Sequence[Any]]:
    """Generate all Gellermann series of length n in lexicographic order.

    Parameters
    ----------
    n
        The length of the series.
    choices
        The two elements of the series (default: ('A', 'B')).
    kwargs
        Additional keyword arguments passed to `is_gellermann_series`.

    Yields
    ------
    Iterator[Sequence[Any]]
        A generator object with all Gellermann series of length n.
    """
    for s in generate_all_boolean_gellermann_series(n, **kwargs):
        yield [choices[int(x)] for x in s]


def _series_to_wide_format_df(series_list: List[Sequence[Any]]) -> pd.DataFrame:
    """Convert a list of series to a wide format DataFrame."""
    series_dicts = [{'series_i': i, **{f'element_{j}': x for j, x in enumerate(s)}} for i, s in enumerate(series_list)]
    return pd.DataFrame(series_dicts).set_index('series_i')


def _series_to_long_format_df(series_list: List[Sequence[Any]]) -> pd.DataFrame:
    """Convert a list of series to a long format DataFrame."""
    series_dicts = [{'series_i': i, 'element_i': j, 'element': e} for i, s in enumerate(series_list) for j, e in enumerate(s)]
    return pd.DataFrame(series_dicts).set_index(['series_i', 'element_i'])


def generate_gellermann_series_table(n: int, m: int, long_format: bool = False, **kwargs: Any) -> pd.DataFrame:
    """Generate a Pandas DataFrame of m random Gellermann series of length n.

    In the wide format, the DataFrame has columns  series_i', 'element_0', 'element_1', ...,
    'element_{n-1}', and each row contains a full series. In the long format, the DataFrame has
    columns 'series_i', 'element_i', 'element', and each row contains a single element of a series.

    Parameters
    ----------
    n
        The length of the series.
    m
        The number of series to generate.
    long_format
        If True, the DataFrame is in long format (default: False).
    kwargs
        Additional keyword arguments passed to `generate_gellermann_series`.

    Returns
    -------
    pd.DataFrame
        A Pandas DataFrame of m random Gellermann series of length n.
    """
    assert n % 2 == 0
    assert m > 0

    generated_series = list(generate_gellermann_series(n, m, **kwargs))
    if long_format:
        return _series_to_long_format_df(generated_series)
    else:
        return _series_to_wide_format_df(generated_series)
